---
source: Rmd
title: "Graphics with ggplot2"
teaching: XX
exercises: XX
questions:
- "How can I generate plots from tabular data in R?"
objectives:
- Make scatter plots and box plots to display your data.
- Use aesthetics to change how variables map to axes, color, or size.
keypoints:
- "Although other plotting systems exist in R, ggplot2 is a good choice for beginners because it is very powerful and relatively user-friendly."
- "Different components of the plot can be added together, much like layers in a Photoshop document."
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
```

## Setup

### Dataset for this lesson

If you don't already have it from the previous lesson, download the dataset from
[Blackmore *et al.* (2017)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5544260/)
and read it into R:

```{r, eval = TRUE}
if (!file.exists("data/rnaseq.csv"))
    download.file(url = "https://raw.githubusercontent.com/UCLouvain-CBIO/WSBIM1207/scrnadata/data/rnaseq.csv",
                  destfile = "data/rnaseq.csv")

rna <- read.csv("data/rnaseq.csv")

head(rna)
```

### Install and load ggplot2

The default R installation does not include ggplot2, so you will need to install
it if you haven't already:

```{r, eval = FALSE}
install.packages("ggplot2")
```

And then you'll load it with `library` in order to use it:

```{r, eval = TRUE}
library(ggplot2)
```

## Getting help

The [website for ggplot2](https://ggplot2.tidyverse.org/) has a cheatsheet and
other useful documentation.  This is a very popular R package, so Google and
Stack Overflow are also very helpful for getting answers!

## Your first plots

The base R installation has the `plot` function and some other functions
for generating plots.  You can make a scatter plot or box plot pretty
quickly this way, but it might not look very nice.

> ## Data transformation
> 
> Because the gene expression data are not normally distributed, here we are
> using the `log1p` function to log transform them.  Because the log of zero
> is negative infinity, and because we have zeros in our dataset, `log1p`
> adds one before log transforming.
{: .callout}

```{r ggfig010, eval = TRUE}
plot(x = rna$time, y = log1p(rna$expression))
```

```{r ggfit020, eval = TRUE}
plot(x = as.factor(rna$infection), y = log1p(rna$expression))
```

I can make the same plots with ggplot2, and hopefully both the code and the
graphics will look a bit better.  (Note that although the package is called
ggplot2, the function is `ggplot`.)

```{r ggfig030, eval = TRUE}
ggplot(rna, aes(x = time, y = log1p(expression))) +
  geom_point()
```

```{r ggfig040, eval = TRUE}
ggplot(rna, aes(x = infection, y = log1p(expression))) +
  geom_boxplot()
```

## Components of a ggplot command

### Call to `ggplot`

In both of the above plots, you'll notice that the first command we used was
`ggplot`.  If you look at the help page for that function, you'll see that there
are two arguments (or, at least, two arguments that can be used): `data` and
`mapping`.  We gave the `data` argument our data frame.  What does this look like
by itself?

```{r ggfig050, eval = TRUE}
ggplot(rna)
```

It basically opened an empty plot, because we gave it data but didn't say anything
about what to do with it.

To the `mapping` argument, we supplied an object that we created using the `aes`
function.  The `aes` function does not have any use outside of ggplot2, but within
ggplot2 it indicates how the variables from the data frame should be used in
the plot.  The function knows to look for variable names in the column names of
the data frame that was supplied, which is very helpful.  What do we get when
we add this in?

```{r ggfig060, eval = TRUE}
ggplot(rna, aes(x = infection, y = log1p(expression)))
```

Now the axes are set up because we've said what we're putting onto them, but
we still haven't plotted anything.

### Geoms

Now we need to add one or more "geoms" to the command to plot our data.
Previously we added a box plot.  We could also do a violin plot, which is a hybrid
between a box plot and density plot.

Notice the use of the `+` sign to combine different commands.  That's pretty
different from the syntax of any other R package, but it's how we do things with
ggplot!

```{r ggfig070, eval = TRUE}
ggplot(rna, aes(x = infection, y = log1p(expression))) +
  geom_violin()
```

There can be multiple geoms in one plot.

```{r ggfig080, eval = TRUE}
ggplot(rna, aes(x = infection, y = log1p(expression))) +
  geom_violin() +
  geom_point()
```

> ## Challenge: Changing the geom order
>
> Try switching `geom_point` with `geom_violin` in the above command, so that
> `geom_point` is first and `geom_violin` is second.  What happens?
>
> > ## Solution
> > 
> > Most of the points are no longer visible, because they are now underneath
> > the violin plot.  Geoms are plotted in the order they appear in the command.
> >
> > ```{r ggfig090, eval = TRUE}
> > ggplot(rna, aes(x = infection, y = log1p(expression))) +
> >   geom_point() +
> >   geom_violin()
> > ```
> {: .solution}
{: .challenge}

In these examples, we didn't provide any arguments to the geom functions.
However, they have their own `data` and `mapping` arguments if you want to
override what was provided in the `ggplot` function.

```{r ggfig100, eval = TRUE}
ggplot(rna, aes(x = infection, y = log1p(expression))) +
  geom_violin() +
  geom_point(mapping = aes(y = log1p(expression) + 10))
```

A quick way to see what other geoms are available is to start typing
`geom` into your R console and look at the autocomplete list.

### More aesthetics

If I look at `?geom_violin` and scroll down to the "Aesthetics" section, there
is a list of other aesthetics we can control.  Although I have treatment on the
x-axis, it might help visually to also color by treatment.  Since this only
applies to one geom, I'll only pass it to that one.

```{r ggfig110, eval = TRUE}
ggplot(rna, aes(x = infection, y = log1p(expression))) +
  geom_violin(mapping = aes(fill = infection)) +
  geom_point()
```

To change an aesthethic value across the board, rather than have it indicate the
value of a variable, we can define it outside of the `mapping` argument.  For
example, I can set `alpha` to control point transparency.

```{r ggfig120, eval = TRUE}
ggplot(rna, aes(x = infection, y = log1p(expression))) +
  geom_violin(mapping = aes(fill = infection)) +
  geom_point(alpha = 0.01)
```

### Titles and axis labels

You might want your plot to have a title, or for the axis labels to be something
different from the variable names.  We'll use the `labs` function for that.

```{r ggfig130, eval = TRUE}
ggplot(rna, aes(x = infection, y = log1p(expression))) +
  geom_violin(mapping = aes(fill = infection)) +
  geom_point(alpha = 0.01) +
  labs(x = "Treatment", fill = "Treatment", y = "logCPM",
       title = "Gene expression vs. viral infection")
```

### Facets

When you are working with a lot of variables, it might make sense to split the
graph into multiple plots.  We can do that with the `facet_wrap` and `facet_grid`
functions.

```{r ggfig140, eval = TRUE}
ggplot(rna, aes(x = infection, y = log1p(expression))) +
  geom_violin(mapping = aes(fill = infection)) +
  geom_point(alpha = 0.01) +
  labs(x = "Treatment", fill = "Treatment", y = "logCPM",
       title = "Gene expression vs. viral infection") +
  facet_wrap(~ tissue)
```

```{r ggfig150, eval = TRUE}
ggplot(rna, aes(x = infection, y = log1p(expression))) +
  geom_violin(mapping = aes(fill = infection)) +
  geom_point(alpha = 0.01) +
  labs(x = "Treatment", fill = "Treatment", y = "logCPM",
       title = "Gene expression vs. viral infection") +
  facet_grid(sex ~ tissue)
```

### Scales

I log-transformed my gene expression before plotting, but maybe I'd prefer
to see it in the original units.  I can change how my y-axis is set up
with `scale_y_continuous`.  The `trans` argument says how I want to transform
the axis, and the `breaks` argument tells it where to put axis ticks and grid
lines.

```{r ggfig160, eval = TRUE}
ggplot(rna, aes(x = infection, y = expression)) +
  geom_violin(mapping = aes(fill = infection)) +
  geom_point(alpha = 0.01) +
  labs(x = "Treatment", fill = "Treatment", y = "Expression",
       title = "Gene expression vs. viral infection") +
  facet_grid(sex ~ tissue) +
  scale_y_continuous(trans = "log1p",
                     breaks = c(0, 1e2, 1e3, 1e4, 1e5, 1e6))
```

> ## Challenge: Changing the fill colors
>
> I'd like the violin plots to be filled with something other than the default
> salmon and teal.  Look at the help page for `scale_fill_manual` and see if
> you can figure out how to use it.  You can specify colors with simple strings
> like `"green"` and `"blue"`.
> 
> If you have extra time, you can play around with color names from
> [this list](http://www.stat.columbia.edu/~tzheng/files/Rcolor.pdf).
> You can also use a [color picker](https://www.w3schools.com/colors/colors_picker.asp)
> to generate custom color strings like `#4d94ff`
>
> > ## Solution
> >
> > ```{r ggfig170, eval = TRUE}
> > ggplot(rna, aes(x = infection, y = expression)) +
> >  geom_violin(mapping = aes(fill = infection)) +
> >  geom_point(alpha = 0.01) +
> >  labs(x = "Treatment", fill = "Treatment", y = "Expression",
> >       title = "Gene expression vs. viral infection") +
> >  facet_grid(sex ~ tissue) +
> >  scale_y_continuous(trans = "log1p",
> >                     breaks = c(0, 1e2, 1e3, 1e4, 1e5, 1e6)) +
> >  scale_fill_manual(values = c(InfluenzaA = "darkolivegreen4",
> >                               NonInfected = "royalblue"))
> > ```
> {: .solution}
{: .challenge}

### Themes

If you want to change something like grid lines or the angle of axis tick labels,
you probably want to add a `theme` command.  There are some pre-build themes, for
example:

```{r ggfig180, eval = TRUE}
ggplot(rna, aes(x = infection, y = log1p(expression))) +
  geom_violin(mapping = aes(fill = infection)) +
  geom_point(alpha = 0.01) +
  labs(x = "Treatment", fill = "Treatment", y = "logCPM",
       title = "Gene expression vs. viral infection") +
  facet_grid(sex ~ tissue) +
  theme_bw()
```

If I just wanted to change the font:

```{r ggfig190, eval = TRUE}
ggplot(rna, aes(x = infection, y = log1p(expression))) +
  geom_violin(mapping = aes(fill = infection)) +
  geom_point(alpha = 0.01) +
  labs(x = "Treatment", fill = "Treatment", y = "logCPM",
       title = "Gene expression vs. viral infection") +
  facet_grid(sex ~ tissue) +
  theme(text = element_text(family = "serif"))
```

Rotating axis labels:

```{r ggfig200, eval = TRUE}
ggplot(rna, aes(x = infection, y = log1p(expression))) +
  geom_violin(mapping = aes(fill = infection)) +
  geom_point(alpha = 0.01) +
  labs(x = "Treatment", fill = "Treatment", y = "logCPM",
       title = "Gene expression vs. viral infection") +
  facet_grid(sex ~ tissue) +
  theme(text = element_text(family = "serif"),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
```

There are too many options here to unpack in a short lesson, but look
at `?theme` and `?element_text` for more information.

### Coordinate systems

What if I want to rotate my plot 90 degrees?  I can do that with `coord_flip`.

```{r ggfig210, eval = TRUE}
ggplot(rna, aes(x = infection, y = log1p(expression))) +
  geom_violin(mapping = aes(fill = infection)) +
  geom_point(alpha = 0.01) +
  labs(x = "Treatment", fill = "Treatment", y = "logCPM",
       title = "Gene expression vs. viral infection") +
  facet_grid(sex ~ tissue) +
  coord_flip()
```

## Using other tidyverse packages with ggplot2

The tidyverse is a group of R packages that are designed to work together to make
data science easier for everyone.  Here are a couple examples of handy functions
to use with ggplot2.

### dplyr

Say I only wanted to look at expression of the _Swt1_ gene. I could filter the
data frame in the "base R" fashion (note that I am putting `alpha` back to the
default since I have far fewer points now):

```{r ggfig220, eval = TRUE}
ggplot(rna[rna$gene == "Swt1",], aes(x = infection, y = log1p(expression))) +
  geom_violin(mapping = aes(fill = infection)) +
  geom_point() +
  labs(x = "Treatment", fill = "Treatment", y = "logCPM",
       title = "Gene expression vs. viral infection") +
  facet_grid(sex ~ tissue)
```

The [dplyr](https://dplyr.tidyverse.org/) package has a lot of functions for
data frame manipulation, and can be used to make a slightly more readable
version of the above command.  The `filter` function is helpful if you only want
to keep certain rows.  We'll also use the pipe command `%>%` (originally from
the magrittr package) to send output of one expression as input to the next
function so we don't have to nest functions so much.

```{r ggfig230, eval = TRUE}
library(dplyr)

rna %>%
  filter(gene == "Swt1") %>%
  ggplot(mapping = aes(x = infection, y = log1p(expression))) +
  geom_violin(mapping = aes(fill = infection)) +
  geom_point() +
  labs(x = "Treatment", fill = "Treatment", y = "logCPM",
       title = "Gene expression vs. viral infection") +
  facet_grid(sex ~ tissue)
```

### tidyr

The [tidyr](https://raw.githubusercontent.com/rstudio/cheatsheets/master/tidyr.pdf) R package
can help you clean up and arrange your data.  In particular, I often find myself
in a situation where I imported something as multiple columns that needs to be
a single column, or vice versa.  The `pivot_longer` and `pivot_wider` functions
are helpful in those cases.

Say I want to have a separate gene expression column for each sample.

```{r, eval = TRUE}
library(tidyr)

rna_wider <- pivot_wider(rna[,c("gene", "sample", "expression")],
                         names_from = sample, values_from = expression)

head(rna_wider)
```

The output is a `tibble`, which is really just the tidyverse version of a data frame.
Anyway, now we can make a plot like the one below to compare each gene across two samples.

```{r ggfig240, eval = TRUE}
ggplot(rna_wider, aes(x = log1p(GSM2545336), y = log1p(GSM2545337))) +
  geom_point()
```

If I had started with a data frame set up like `rna_wider` and wanted it in long
format, I could do the opposite operation using `pivot_longer`.

> ## Selecting columns
>
> See `?tidyr_tidy_select` for methods to select columns in the tidyverse,
> including `starts_with`.
{: .callout}

```{r, eval = TRUE}
rna_longer <- pivot_longer(rna_wider, cols = starts_with("GSM"),
                           names_to = "sample", values_to = "expression")

head(rna_longer)
```

All the sample metadata are missing from `rna_longer`, however.  I'll build a
metadata table here from the original, although you could have hypothetically
imported such metadata from a CSV.

```{r, eval = TRUE}
metadata <- rna[,c("sample", "sex", "infection", "time", "tissue", "mouse")]
metadata <- metadata[!duplicated(metadata),]
head(metadata)
```

We could then use `left_join` from the `dplyr` package to recreate the original
dataset.

```{r ggfig250, eval = TRUE}
rna_orig <- left_join(rna_longer, metadata, by = "sample")
head(rna_orig)

ggplot(rna_orig, aes(x = infection, y = log1p(expression))) +
  geom_violin(mapping = aes(fill = infection)) +
  geom_point(alpha = 0.01) +
  labs(x = "Treatment", fill = "Treatment", y = "logCPM",
       title = "Gene expression vs. viral infection") +
  facet_grid(sex ~ tissue)
```

